plugin/src/test/java/org/wildfly/plugins/bootablejar/maven/goals/DevWatchContextTestCase.java

```diff
@@ -0,0 +1,726 @@
+                }
+            });
+            expectedTargetFile = targetDir.resolve(existingFile.getFileName());
+            Assert.assertEquals(1, handler.deleted.size());
+            Assert.assertEquals(expectedTargetFile, handler.deleted.get(existingFile));
+            Assert.assertEquals(1, handler.copied.size());
+            Assert.assertEquals(expectedTargetFile, handler.copied.get(existingFile));
+            Assert.assertEquals("Hello4", new String(Files.readAllBytes(expectedTargetFile)));
+            projCtx.reset();
+
+            // Delete the existing file
+            Files.delete(existingFile);
+            handler = checkEvent(ctx, projCtx, false, false, false, false, false, new Checker() {
+                @Override
+                public boolean isOk(BootableAppEventHandler handler) {
+                    return handler.deleted.size() == 1;
+                }
+            });
+            expectedTargetFile = targetDir.resolve(existingFile.getFileName());
+            Assert.assertEquals(1, handler.deleted.size());
+            Assert.assertEquals(expectedTargetFile, handler.deleted.get(existingFile));
+            Assert.assertFalse(Files.exists(expectedTargetFile));
+            projCtx.reset();
+
+            //Create a new directory and sub directory
+            Path newDir1 = project.webAppDir.resolve("foo");
+            Path newDir2 = newDir1.resolve("bar");
+            Files.createDirectories(newDir2);
+            handler = checkEvent(ctx, projCtx, false, false, false, false, false, new Checker() {
+                @Override
+                public boolean isOk(BootableAppEventHandler handler) {
+                    return handler.monitored.size() == 2;
+                }
+            });
+            Assert.assertEquals(2, handler.monitored.size());
+            Assert.assertTrue(handler.monitored.contains(newDir1));
+            Assert.assertTrue(handler.monitored.toString(), handler.monitored.contains(newDir2));
+            projCtx.reset();
+
+            // Add a new file in the new sub directory
+            newFile = Files.createFile(newDir2.resolve("index3.html"));
+            expectedTargetFile = targetDir.resolve(project.webAppDir.relativize(newFile));
+            Files.write(newFile, "Hello3".getBytes());
+            handler = checkEvent(ctx, projCtx, false, false, false, false, false, new Checker() {
+                @Override
+                public boolean isOk(BootableAppEventHandler handler) {
+                    return handler.copied.size() == 1;
+                }
+            });
+            Assert.assertEquals(1, handler.copied.size());
+            Assert.assertEquals(handler.copied.toString(), expectedTargetFile, handler.copied.get(newFile));
+            Assert.assertEquals("Hello3", new String(Files.readAllBytes(expectedTargetFile)));
+            projCtx.reset();
+
+            // Delete directory, random failure on Windows.
+            if (!isWindows()) {
```

Comment:
Do we suspect this is a real problem? Maybe something to do with the previous `DevWatchContext.cleanUp()` method.

Reply:
I don't think so. I think that this in timing issue on Windows. The DevWatchContext.cleanUp() does nothing in unit tests. It just set a flag that it has been called.  We are not touching at the file system (other than what the test does).